plane/src/controller/proxy.rs

```diff
@@ -19,31 +22,135 @@ use std::net::{IpAddr, SocketAddr};
 use tokio::select;
 use valuable::Valuable;
 
-pub async fn handle_message_from_proxy(
-    message: MessageFromProxy,
+pub async fn handle_route_info_request(
+    token: BearerToken,
     controller: &Controller,
     socket: &mut TypedSocket<MessageToProxy>,
-    cluster: &ClusterName,
-    node_id: NodeId,
 ) {
-    match message {
-        MessageFromProxy::RouteInfoRequest(RouteInfoRequest { token }) => {
-            let route_info = match controller.db.backend().route_info_for_token(&token).await {
-                Ok(route_info) => route_info,
-                Err(err) => {
-                    tracing::error!(?err, "Error getting route info");
-                    return;
-                }
+    match controller.db.backend().route_info_for_token(&token).await {
+        // When a proxy requests a route, either:
+        // 1. The route is ready, and we can send it back immediately.
+        // 2. The route is not ready, and we need to wait for it to become ready.
+        // 3. The route does not exist or has already terminated, and we can send back a `None`.
+        Ok(RouteInfoResult::Ready(route_info)) => {
+            let response = RouteInfoResponse {
+                token,
+                route_info: Some(route_info),
             };
+            if let Err(err) = socket
+                .send(MessageToProxy::RouteInfoResponse(response))
+                .await
+            {
+                tracing::error!(?err, "Error sending route info response to proxy.");
+            }
+        }
+        Ok(RouteInfoResult::Waiting(partial_route_info)) => {
+            let backend_id = partial_route_info.backend_id.clone();
+            let mut sub: Subscription<BackendState> =
+                controller.db.subscribe_with_key(backend_id.as_str());
+
+            // There is a race condition if the status updated between when our last query hit and when we started the
+            // subscription. It's a bit hacky, but for now we will just issue the query again.
+            // We can't start the subscription first to avoid repeating the query, because we need to know the backend
+            // ID to start the subscription.
+            if let Ok(RouteInfoResult::Ready(route_info)) =
+                controller.db.backend().route_info_for_token(&token).await
+            {
```

Comment:
It's also possible for the backend to have terminated in which case we should also drop out.

Reply:
This will also catch that case (`ready` here means that the RouteInfoResult is ready, not that the backend itself is, so if the backend is terminated a "terminated" RouteInfoResult will be ready). The overlap with backend states is confusing, so I renamed Waiting/Ready it to Pending/Available.

Reply:
It looks like `route_info_for_token()` returns `Ok(RouteInfoResult::NotFound)` if the backend is terminated or terminating, though.

Reply:
Ah good catch, fixed.